subprojects/hashing/src/main/java/org/gradle/internal/hash/HashCode.java

```diff
@@ -95,98 +109,229 @@ private static int decode(char ch) {
+    public abstract int length();
+
+    public abstract byte[] toByteArray();
+
+    @Override
+    public abstract int hashCode();
+
+    @Override
+    public abstract boolean equals(@Nullable Object obj);
+
+    @Override
+    public String toString() {
+        StringBuilder sb = toStringBuilder(2 * length(), bytes());
+        return sb.toString();
     }
 
-    public byte[] toByteArray() {
-        return bytes.clone();
+    public String toZeroPaddedString(int length) {
+        StringBuilder sb = toStringBuilder(length, bytes());
+        while (sb.length() < length) {
+            sb.insert(0, '0');
+        }
+        return sb.toString();
     }
 
-    @Override
-    public int hashCode() {
-        if (hashCode == 0) {
-            hashCode = (bytes[0] & 0xFF)
-                | ((bytes[1] & 0xFF) << 8)
-                | ((bytes[2] & 0xFF) << 16)
-                | ((bytes[3] & 0xFF) << 24)
-                // Make sure it's always > 0 but without affecting the lower 32 bits
-                | (1L << 32);
+    private static StringBuilder toStringBuilder(int capacity, byte[] bytes) {
+        StringBuilder sb = new StringBuilder(capacity);
+        for (byte b : bytes) {
+            sb.append(HEX_DIGITS[(b >> 4) & 0xf]).append(HEX_DIGITS[b & 0xf]);
         }
-        return (int) hashCode;
+        return sb;
     }
 
-    @Override
-    public boolean equals(@Nullable Object obj) {
-        if (obj == this) {
-            return true;
+    public String toCompactString() {
+        return new BigInteger(1, bytes()).toString(36);
+    }
+
+    // Package private accessor used by MessageDigestHasher.putHash()
+    abstract void appendToHasher(PrimitiveHasher hasher);
+
+    abstract byte[] bytes();
+
+    static class HashCode128 extends HashCode {
+        private final long bits1;
+        private final long bits2;
+
+        public HashCode128(long bits1, long bits2) {
+            this.bits1 = bits1;
+            this.bits2 = bits2;
         }
 
-        if (obj == null || obj.getClass() != HashCode.class) {
-            return false;
+        @Override
+        public int length() {
+            return 16;
         }
 
-        byte[] a = bytes;
-        byte[] b = ((HashCode) obj).bytes;
-        int length = a.length;
+        @Override
+        byte[] bytes() {
+            return toByteArray();
+        }
 
-        if (b.length != length) {
-            return false;
+        @Override
+        public byte[] toByteArray() {
+            byte[] bytes = new byte[16];
+            longToBytes(bits1, bytes, 0);
+            longToBytes(bits2, bytes, 8);
+            return bytes;
         }
 
-        for (int i = 0; i < length; i++) {
-            if (a[i] != b[i]) {
+        @Override
+        void appendToHasher(PrimitiveHasher hasher) {
+            hasher.putLong(bits1);
+            hasher.putLong(bits2);
+        }
+
+        @Override
+        public int hashCode() {
+            return (int) bits1;
```

Comment:
Is that a good hashCode to use? Java uses by default something like `(int)(value ^ (value >>> 32))` for long (according to Long.hashCode() code). 

Just wondering if there is any case that there would be some advantage of doing this here.

Reply:
As far as I understand, the bits come from MD5 or a similar hash function, so any 32 bit fragment should have good distribution. That is why just a single bits field is enough

I think it is worth clarifying in a code comment since it does surprise readers

Reply:
That's the reason, yes. Same reason why we also used to calculate `HashCode.hashCode()` by combining only the first four bytes.

I don't think we need a comment here, as if you are touching this code, you should know how cryptographic hashes work.

Reply:
I was thinking from the perspective that `HashCode` can be created with any random byte input. So implementation relies on the fact that input has good distribution. But I get the point, that should not be the case here.